### DISCUSSION TITLE
2-loop

### DISCUSSION THREAD

**Discussion by MAR-TP** [FeynCalc/feyncalc#138] [Q&A]
Hello, I am student and  I am using Windows 7 x64;  FeynCalc 9.3.1;  FeynArts 3.11 for my calculations. 

For calculations with 1-loop I used the command M3 = (TID [# 1, q, ToPaVe -> True] &) [SUNSimplify [M2]] and after this command I used MsqAvg1 =
  Simplify [(TrickMandelstam [# 1, {s, t, u,
        2 * m ^ 2 + mg ^ 2}] &) [(DoPolarizationSums [# 1, k2,
        k1] &) [(DoPolarizationSums [# 1, k1, 0, VirtualBoson -> True,
         GaugeTrickN -> 4] &) [
      DiracSimplify [(FermionSpinSum [# 1,
           ExtraFactor -> 1/2 ^ 2] &) [(SUNSimplify [# 1, Explicit -> True,
            SUNNToCACF -> False] &) [
         FeynAmpDenominatorExplicit [(1 / SUNN ^ 2) * (M4 *
             ComplexConjugate [M4])]]]]]]]] to calculate the squared amplitude. 

At this point I need to do some calculations for a 2-loop where q1 and q2 are the loop pulses.  I used the command M3 = (FCMultiLoopTID [M2, {q1, q2}] &) [SUNSimplify [M2]] and wrote the above command MsqAvg1 after it, but the loop pulses q1 and q2 remain in the expressions.  

Also my computer is old and it takes a long time to compute.  Can you tell me which command to use to get rid of the loop impulses q1 and q2 and calculate squared amplitude.  Thank you in advance for your response.

![photo_2021-10-07_17-07-28](https://user-images.githubusercontent.com/92045591/136391019-e72b4622-7f6b-43b5-86f5-3b291cfb45aa.jpg)

**Comment by vsht** [comment]
I'm not sure how to answer this question, so I perhaps I'll ask first if you did any 2-loop calculation before?

`FCMultiLoopTID` will merely do tensor reduction w.r.t loop momenta contracted with Dirac matrices, epsilon tensors and polarization vectors. But of course you still have loop momenta contracted to other loop momenta and external momenta. You can't get rid of those
beyond 1-loop doing only tensor reductions.

And then you also need to do an IBP-reduction, get the list of the final master integrals, calculate them
and plug that back into the expression. These steps are not part of FeynCalc and anyway the expressions one usually has to
deal with are too large for Mathematica. For that there are FORM (amplitude manipulation) and FIRE/KIRA/LiteRed/... (IBP reduction).
Actually, even intermediate expressions from the multi-loop tensor reduction alone (what FCMultiLoopTID does) can already
become so large that things become painfully slow. This is why it takes so long to compute. Already for this step one should go
with FORM. You can still get the tensor reduction rules from FeynCalc, like 
```
Tdec[{{q1,mu},{q2,nu}},{p1,p2},List->False]
```
for a rank 2 two-loop tensor integral depending on 2 external momenta that you then export as  a FORM id statement.

So yeah, computing anything beyond 1-loop usually involves FORM, lots of custom codes and interfacing with other programs.
You can't just do everything in a Mathematica notebook with few commands, except for perhaps few exceptionally simple cases.

**Reply by MAR-TP** [reply]
Thank you for your advice on how to calculate diagrams with a 2-loop.  I also got acquainted with the example: https://feyncalc.github.io/FeynCalcExamplesMD/QCD/TwoLoops/Gh-Gh

Since I haven't done calculations with 2-loop before, can you send me some additional examples with FeynCalc and Form?

Thank you in advance.

**Comment by vsht** [comment]
Sorry, I don't really have any more worked out examples apart from what is available in the example gallery.
For time reasons alone, the amount of effort I can put into FeynCalc support is very limited. I'm not getting
paid of that and even academically this is not counted as research, rather like hobby programming.

In its current state FeynCalc is not really useful for stuff beyond 1-loop, so you can in principle forget it.
The single 2-loop example you mentioned is more like an exception but not something you can build upon.

FORM codes for loop calculations are usually hand-written for a particular processes and kept private
for obvious reasons (you don't want other people to calculate and publish stuff you are currently
working on yourself). Groups that do such calculations regularly usually have in-house codes but
again this is not something one would share or make public.

Usually you learn that kind of stuff when you embark on a particular project. So I would talk to
people with whom you are collaborating on this calculation. Learning everything from scratch
alone is pretty tough but it often happens that towards the end of the project people have
their own frameworks ready.

You can, however, have a look at

https://indico.desy.de/event/26814/timetable/?view=nicecompact

and in particular

https://indico.desy.de/event/26814/contributions/94605/attachments/65148/80414/magerya-capp2021.pdf
https://github.com/magv/alibrary

I think it is one of the very few more advanced codes publicly available. Also the slides are quite informative. 
But if you are not familiar with FORM there is still a long way to go...
